packages/flutter/lib/src/material/text_field.dart

```diff
@@ -207,6 +251,26 @@ class _TextFieldState extends State<TextField> {
   FocusNode _focusNode;
   FocusNode get _effectiveFocusNode => widget.focusNode ?? (_focusNode ??= new FocusNode());
 
+  bool get needsCounter => widget.maxLength != null
+    && widget.decoration != null
+    && widget.decoration.counterText == null;
+
+  InputDecoration get _effectiveDecoration {
+    if (!needsCounter)
+      return widget.decoration;
+
+    final String counterText = '${_effectiveController.value.text.length} / ${widget.maxLength}';
+    if (_effectiveController.value.text.length > widget.maxLength) {
```

Comment:
This doesn't count characters. It counts UTF-16 code units, which isn't even remotely close to what users think of as characters.

If you want to count characters, you should count extended grapheme clusters.

Reply:
I've implemented it as far as using Unicode "runes" in Dart.  In order to do better than that, [this Dart feature](https://github.com/dart-lang/sdk/issues/28404) needs to be implemented, or we need to build out an implementation in the engine that can use ICU to do these kinds of operations for us.  That's highly non-trivial, so punting that with some TODOs and user documentation.